### PR TITLE
services: reflect filters also in the URL

### LINKED ISSUE #16788: services: Services page should show filters in the URL, so it can be sharable and bookmarkable
When changing filters on the services page, it should adjust the URL with the filter options. This would enable someone to share the link with others or bookmark it to have a custom view on their services.

### PR DESCRIPTION
**Pull request by KKoukiou**
Fixes #16788

### CODE REVIEW

File: pkg/systemd/services/services.jsx [hunk 1]
```diff
@@ -926,19 +956,16 @@ const ServicesPageFilters = ({
         onDeleteChip();
     }, [setCurrentTextFilter, onDeleteChip]);
 
+    const onTextFilterChanged = textFilter => {
+        setCurrentTextFilter(textFilter);
+        cockpit.location.go([], ({ ...options, name: textFilter }));
+    };
```
Comment (cockpituous) [inline pkg/systemd/services/services.jsx:962]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18115-20230103-122827-2f1a70ab-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/services/services.jsx [hunk 2]
```diff
@@ -947,8 +974,8 @@
                              className="services-text-filter"
                              placeholder={_("Filter by name or description")}
                              value={currentTextFilter}
-                             onChange={setCurrentTextFilter}
-                             onClear={() => setCurrentTextFilter('')} />
+                             onChange={onTextFilterChanged}
+                             onClear={() => onTextFilterChanged('')} />
```
Comment (cockpituous) [inline pkg/systemd/services/services.jsx:978]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18115-20230103-122827-2f1a70ab-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/services/services.jsx [hunk 3]
```diff
@@ -947,8 +971,8 @@ const ServicesPageFilters = ({
...
```
Comment (cockpituous) [inline pkg/systemd/services/services.jsx:975]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18115-20230103-144537-e584facb-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/services/services.jsx [hunk 4]
```diff
@@ -926,19 +949,15 @@ const ServicesPageFilters = ({
         onDeleteChip();
     }, [setCurrentTextFilter, onDeleteChip]);
 
+    const onTextFilterChanged = textFilter => {
+        setCurrentTextFilter(textFilter);
+    };
```
Comment (cockpituous) [inline pkg/systemd/services/services.jsx:954]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18115-20230104-081322-a77aca1f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/services/services.jsx [hunk 5]
```diff
@@ -947,8 +966,8 @@
                              className="services-text-filter"
                              placeholder={_("Filter by name or description")}
                              value={currentTextFilter}
-                             onChange={setCurrentTextFilter}
-                             onClear={() => setCurrentTextFilter('')} />
+                             onChange={onTextFilterChanged}
+                             onClear={() => onTextFilterChanged('')} />
```
Comment (cockpituous) [inline pkg/systemd/services/services.jsx:970]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18115-20230104-081322-a77aca1f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/services/services.jsx [hunk 6]
```diff
@@ -947,8 +976,8 @@ const ServicesPageFilters = ({
...
```
Comment (cockpituous) [inline pkg/systemd/services/services.jsx:980]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18115-20230104-115006-dfe9a654-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/services/services.jsx [hunk 7]
```diff
@@ -870,17 +885,35 @@ const ServicesPageFilters = ({
...
+
+        setCurrentTextFilter(decodeURIComponent(options.name || ""));
+        setFilters(_filters);
+    }, [options, setCurrentTextFilter, setFilters]);
+
+    useEffect(() => {
+        const _options = {};
+
+        _options.activestate = JSON.stringify(filters.activeState);
+        _options.filestate = JSON.stringify(filters.fileState);
```
Comment (martinpitt) [inline pkg/systemd/services/services.jsx:911]: That JSON stringify/parse dance rings a bell :grimacing:  Still terrible, though.. (No action, just thinking aloud)

File: pkg/systemd/services/services.jsx (hunk 6 above)
Comment (martinpitt) [reply pkg/systemd/services/services.jsx:980]: That's a good point -- we never test clearing the filter?

**Review by martinpitt — APPROVED** [review]
Thanks! Looks good, for bonus points please add a check that clearing the filter actually works. (Or is our coverage check broken?)

File: pkg/systemd/services/services.jsx (hunk 6 above)
Comment (KKoukiou) [reply pkg/systemd/services/services.jsx:980]: I will add a test as a followup.

File: pkg/systemd/services/services.jsx (hunk 7 above)
Comment (KKoukiou) [reply pkg/systemd/services/services.jsx:911]: That's for different reason here :) It's array to URL options. Thought it's enough to encode with json. WDYT?

**Pull request merged**